dockstore-webservice/src/main/java/io/dockstore/webservice/core/Doi.java

```diff
@@ -0,0 +1,145 @@
+
+@Entity
+@Schema(description = "A Digital Object Identifier (DOI)")
+@Table(name = "doi", uniqueConstraints = @UniqueConstraint(name = "unique_doi_name", columnNames = { "name" }))
+public class Doi {
+    // DOI order of precedence from greatest to least
+    public static final List<DoiCreator> DOI_ORDER_OF_PRECEDENCE = List.of(DoiCreator.USER, DoiCreator.GITHUB, DoiCreator.DOCKSTORE);
+    public static final int MAX_NUMBER_OF_DOI_CREATORS = 3;
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Schema(description = "Implementation specific ID for the DOI in this web service")
+    private long id;
+
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    @Schema(description = "The type of DOI", requiredMode = RequiredMode.REQUIRED, example = "CONCEPT")
+    private DoiType type;
+
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    @Schema(description = "The DOI creator", requiredMode = RequiredMode.REQUIRED, example = "USER")
+    private DoiCreator creator;
+
+    @Column(nullable = false)
+    @Schema(description = "The DOI name", requiredMode = RequiredMode.REQUIRED, example = "10.5281/zenodo.705645")
+    private String name;
+
+    @Column
+    @Schema(description = "The ID of the access link with edit permissions")
+    private String editAccessLinkId;
+
+    // database timestamps
+    @Column(updatable = false)
+    @CreationTimestamp
+    private Timestamp dbCreateDate;
+
+    @Column()
+    @UpdateTimestamp
+    private Timestamp dbUpdateDate;
+
+    public Doi() {
+    }
+
+    public Doi(DoiType type, DoiCreator creator, String name) {
+        this.type = type;
+        this.creator = creator;
+        this.name = name;
+    }
+
+    public long getId() {
+        return id;
+    }
+
+    public void setId(long id) {
+        this.id = id;
+    }
+
+    public DoiType getType() {
+        return type;
+    }
+
+    public void setType(DoiType type) {
+        this.type = type;
+    }
+
+    public DoiCreator getCreator() {
+        return creator;
+    }
+
+    public void setCreator(DoiCreator creator) {
+        this.creator = creator;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String getEditAccessLinkId() {
+        return editAccessLinkId;
+    }
+
+    public void setEditAccessLinkId(String editAccessLinkId) {
+        this.editAccessLinkId = editAccessLinkId;
+    }
+
+    public static Doi getDoiBasedOnOrderOfPrecedence(Map<DoiCreator, Doi> dois) {
+        for (DoiCreator doiCreator: DOI_ORDER_OF_PRECEDENCE) {
+            if (dois.containsKey(doiCreator)) {
+                return dois.get(doiCreator);
+            }
+        }
+        return null;
+    }
+
+    public enum DoiType {
+        CONCEPT,
+        VERSION
+    }
+
+    public enum DoiCreator {
+        USER,
+        DOCKSTORE,
+        GITHUB
```

Comment:
So, really nit-picking, and I can't think of a good alternative off-hand, but the DOI is being created by the user's Zenodo account, not by GitHub. The Zenodo/GitHub integration grants GitHub permissions to the Zenodo account, which in turn creates the DOIs. GitHub is not the DOI creator.

I have no alternative proposals off-hand. This could be solved with a comment, and/or we can figure out later how we want to display it in the UI, if we ever surface the creator there.

Reply:
Hmm, how about `DoiInitiator`? The DOI creation in the Zenodo/GitHub integration is initiated by GitHub. Similarly, the DOI creation is initiated by `USER` or `DOCKSTORE` for user-generated and Dockstore-generated DOIs.

Reply:
I think that works. I'd argue that GitHub is the trigger, not the initiator, but it's not worth splitting hairs over that. :)